OSSM 3.0 TP1: OSSM-6060 OSSM3 with cert-manager and istio-csr (#83665)
openshift/openshift-docs: modules/ossm-cert-manager-update-istio-csr-revisionbased-only.adoc

= Updating istio-csr if using RevisionBased update strategy

If you set `spec.updateStrategy` to `RevisionBased` when creating your `Istio` resource, then `istio-csr` requires you to pass all revisions each time you upgrade your `RevisionBased` control plane. To do this, you need to follow this order each time you upgrade:

. Update `istio-csr` deployment with new revision.
. Update `Istio.spec.version`.

[id="example-update-revision-based-control-plane_{context}"]
== Example update for RevisionBased control plane

In this example, the `controlplane` is being updated from `v1.23.0` to `1.23.1.`

. Update `istio-csr` deployment with the new revision:
+
[source, terminal]
----
helm upgrade cert-manager-istio-csr jetstack/cert-manager-istio-csr \
  --install \
  --namespace cert-manager \
  --wait \
  --reuse-values \
  --set "app.istio.revisions={default-v1-23-0,default-v1-23-1}"
----

. Remove the revision from your `istio-csr` deployment:
+
[source, terminal]
----
helm upgrade cert-manager-istio-csr jetstack/cert-manager-istio-csr \
  --install \
  --namespace cert-manager \
  --wait \
  --reuse-values \
  --set "app.istio.revisions={default-v1-23-1}"

> ----

Reviewer: Sorry for the confusion. You still need to update your `Istio.spec.version = v1.23.1` in between adding the new revision and removing the old one.

It should be something like:
1. Update `istio-csr` deployment.
2. Update `Istio.spec.version.
3. Remove the old revision from your `istio-csr` deployment.

In the upstream docs for 2. we simply have
> `Then you can update your `Istio.spec.version = v1.23.1`

but here we may need to provide a whole yaml example?
```
kind: Istio
metadata:
  name: default
spec:
  version: v1.23.1
```

= Updating istio-csr if using RevisionBased update strategy

If you set `spec.updateStrategy` to `RevisionBased` when creating your `Istio` resource, then `istio-csr` requires you to pass all revisions each time you upgrade your `RevisionBased` control plane. To do this, you need to follow this order each time you upgrade:

. Update `istio-csr` deployment with new revision.
. Update `Istio.spec.version`.

[id="example-update-revision-based-control-plane_{context}"]
== Example update for RevisionBased control plane

In this example, the `controlplane` is being updated from `v1.23.0` to `1.23.1.`

. Update `istio-csr` deployment with the new revision:
+
[source, terminal]
----
helm upgrade cert-manager-istio-csr jetstack/cert-manager-istio-csr \
  --install \
  --namespace cert-manager \
  --wait \
  --reuse-values \
  --set "app.istio.revisions={default-v1-23-0,default-v1-23-1}"
----

. Update `Istio.spec.version = v1.23.1`:
+
[source, terminal]
----
helm upgrade cert-manager-istio-csr jetstack/cert-manager-istio-csr \
  --install \
  --namespace cert-manager \
  --wait \
  --reuse-values \
  --set "app.istio.revisions={default-v1-23-1}"

> ----

Reviewer: This says that we're updating the `Istio` resource but the example shows updating the `istio-csr` resource. This example should go under the last step: "Once the old revision is not longer in use ..."

// Module included in the following assemblies:
//
// * service-mesh-docs-main/install/ossm-cert-manager-assembly.adoc

:_mod-docs-content-type: PROCEDURE
[id="updating-istio-csr-revision-based-only_{context}"]

> = Updating istio-csr for RevisionBased update strategy

Reviewer 2: ```suggestion
= Updating istio-csr agents with revision-based update strategies
```

> = Updating istio-csr if using RevisionBased update strategy

Reviewer 3: ```suggestion
= Updating istio-csr for RevisionBased update strategy
```

> If you set `spec.updateStrategy` to `RevisionBased` when creating your `Istio` resource, then `istio-csr` requires you to pass all revisions each time you upgrade your `RevisionBased` control plane. To do this, you need to follow this order each time you upgrade:

Reviewer 3: ```suggestion
When you created your `Istio` resource, if you set the `spec.updateStrategy` field to `RevisionBased`, the `istio-csr` agent requires that you pass all revisions each time you upgrade your `RevisionBased` control plane. To do so, you must follow this order each time you upgrade:
```

> When you created your `Istio` resource, if you set the `spec.updateStrategy` field to `RevisionBased`, the `istio-csr` agent requires that you pass all revisions each time you upgrade your `RevisionBased` control plane. To do so, you must follow this order each time you upgrade:

Reviewer 2: ```suggestion
If you deployed your Istio resource using the revision-based update strategy, you must pass all revisions each time you update your control plane. You must perform the update in the following order:
```
See the glossary for guidance on [update vs upgrade.](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/term_glossary.adoc#update)

> . Update `istio-csr` deployment with new revision.

Reviewer 3: ```suggestion
. Update the `istio-csr` deployment with the new revision.
```

> . Update the `istio-csr` deployment with new revision.

Reviewer 2: ```suggestion
. Update the `istio-csr` deployment with the new revision.
```

> . Update `Istio.spec.version`.

Reviewer 3: ```suggestion
. Update the `Istio.spec.version` value.
```

> . Update the `Istio.spec.version` value.

Reviewer 2: ```suggestion
. Update the value of `Istio.spec.version` parameter/field.
```

[id="example-update-revision-based-control-plane_{context}"]

> == Example update for RevisionBased control plane

Reviewer 2: This isn't formatted as a proper procedure module. And this really doesn't need a second heading. 

You could refactor the commands using variables and a combination of code callouts and a definition list.

In this example, the `controlplane` is being updated from `v1.23.0` to `1.23.1.`

> . Update `istio-csr` deployment with the new revision:

Reviewer 3: ```suggestion
. Update the `istio-csr` deployment with the new revision by running the following command:
```

+
[source, terminal]
----
helm upgrade cert-manager-istio-csr jetstack/cert-manager-istio-csr \
  --install \
  --namespace cert-manager \
  --wait \
  --reuse-values \
  --set "app.istio.revisions={default-v1-23-0,default-v1-23-1}"
----

> . Update `istio.spec.version.` in the `Istio` object:

Reviewer 3: ```suggestion
. Update the `istio.spec.version` value in the `Istio` object by editing the YAML file:
```
I'm not sure if `istio.spec.version` should include two dots or three.

> helm upgrade cert-manager-istio-csr jetstack/cert-manager-istio-csr \
>   --install \
>   --namespace cert-manager \
>   --wait \
>   --reuse-values \
>   --set "app.istio.revisions={default-v1-23-0,default-v1-23-1}"
> ----

Reviewer 2: ```suggestion
$ helm upgrade cert-manager-istio-csr jetstack/cert-manager-istio-csr \
  --install \
  --namespace cert-manager \
  --wait \
  --reuse-values \
  --set "app.istio.revisions={<old_revision>,<new_revision>}"
----
where:
`old_revision` :: Specifies the old revision in the <istio-name>-v<major_version>-<minor_version>-<patch_version>` format. For example: `default-v1-23-0`.
`new_revision` :: Specfies the new revision in the <istio-name>-v<major_version>-<minor_version>-<patch_version>` format. For example: `default-v1-23-1`.
```

+
.Example `istio.yaml`
[source, yaml]
----
apiVersion: sailoperator.io/v1alpha1
kind: Istio
metadata:
  name: default
spec:
  version: v1.23.1
----

> . Remove the old revision from your `istio-csr` deployment:

Reviewer 3: ```suggestion
. Remove the old revision from your `istio-csr` deployment by running the following command:
```

> . Update the `istio.spec.version` in the `Istio` object:
> +
> .Example `istio.yaml`
> [source, yaml]
> ----
> apiVersion: sailoperator.io/v1alpha1
> kind: Istio
> metadata:
>   name: default
> spec:
>   version: v1.23.1
> ----

Reviewer 2: ```suggestion
. Update the `istio.spec.version` in the `Istio` object similar to the following example:
+
.Example `istio.yaml` file
[source, yaml]
----
apiVersion: sailoperator.io/v1alpha1
kind: Istio
metadata:
  name: default
spec:
  version: <new_revision> # <1>
----
<1> Update to the new revision prefixed with the letter _v_, such as `v1.23.1`
```

+
[source, terminal]
----
helm upgrade cert-manager-istio-csr jetstack/cert-manager-istio-csr \
  --install \
  --namespace cert-manager \
  --wait \
  --reuse-values \

>   --set "app.istio.revisions={default-v1-23-1}"

Reviewer 2: I would refactor this part to resemble the suggestions above.
